serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,288 @@
+	ErrPodNameUnset   = errors.New(`"podName" is unset`)
+	ErrNotInCluster   = errors.New("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined")
+)
+
+// New instantiates a Client. The stopCh is used for exiting retry loops
+// when closed.
+func New(logger hclog.Logger, stopCh <-chan struct{}) (*Client, error) {
+	config, err := inClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		logger: logger,
+		config: config,
+		stopCh: stopCh,
+	}, nil
+}
+
+// Client is a minimal Kubernetes client.
+type Client struct {
+	logger hclog.Logger
+	config *Config
+	stopCh <-chan struct{}
+}
+
+// GetPod gets a pod from the Kubernetes API.
+func (c *Client) GetPod(namespace, podName string) (*Pod, error) {
+	endpoint := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s", namespace, podName)
+	method := http.MethodGet
+
+	// Validate that we received required parameters.
+	if namespace == "" {
+		return nil, ErrNamespaceUnset
+	}
+	if podName == "" {
+		return nil, ErrPodNameUnset
+	}
+
+	req, err := http.NewRequest(method, c.config.Host+endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	pod := &Pod{}
+	if err := c.do(req, pod); err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+// PatchPod updates the pod's tags to the given ones.
+// It does so non-destructively, or in other words, without tearing down
+// the pod.
+func (c *Client) PatchPod(namespace, podName string, patches ...*Patch) error {
+	endpoint := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s", namespace, podName)
+	method := http.MethodPatch
+
+	// Validate that we received required parameters.
+	if namespace == "" {
+		return ErrNamespaceUnset
+	}
+	if podName == "" {
+		return ErrPodNameUnset
+	}
+	if len(patches) == 0 {
+		// No work to perform.
+		return nil
+	}
+
+	var jsonPatches []interface{}
+	for _, patch := range patches {
+		if patch.Operation == Unset {
+			return errors.New("patch operation must be set")
+		}
+		jsonPatches = append(jsonPatches, map[string]interface{}{
+			"op":    patch.Operation.String(),
+			"path":  patch.Path,
+			"value": patch.Value,
+		})
+	}
+	body, err := json.Marshal(jsonPatches)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(method, c.config.Host+endpoint, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json-patch+json")
+	return c.do(req, nil)
+}
+
+// do executes the given request, retrying if necessary.
+func (c *Client) do(req *http.Request, ptrToReturnObj interface{}) error {
+	// Finish setting up a valid request.
+	req.Header.Set("Authorization", "Bearer "+c.config.BearerToken)
+	req.Header.Set("Accept", "application/json")
+	client := cleanhttp.DefaultClient()
+	client.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: c.config.CACertPool,
+		},
+	}
+
+	// Execute and retry the request. This exponential backoff comes
```

Comment:
It would be helpful to comment on the initial/max times here. e.g. it's not obvious if the retry delays will start at 5ms, or 5s...

Reply:
Hopefully the explicitness of the vars passed into the retryable HTTP client solves this, but let me know if it doesn't: https://github.com/hashicorp/vault/pull/8249/commits/b00a0c420174d2557d789ba1e9924aa89e73f51a

Reply:
Marking resolved just to track what I've pushed code for.